Expand the out-of-bounds pointers section. (#3)
CTSRD-CHERI/cheri-c-programming: cheri-c-programming.tex

\note{I feel like this section wants a reference to CHERI Concentrate?}{nwf}

ISO C permits pointers to go only one byte beyond their original
allocation, but real code sometimes constructs transient pointer
values which are further out of bounds.  To support this, capabilities
can hold a range of out-of-bounds addresses while retaining a valid
tag, and CHERI-enabled hardware performs bounds checks only on pointer

> use (i.e. dereference) but not on pointer manipulation.  Dereferencing

Reviewer: To avoid weird spacing:
```suggestion
use (i.e.\ dereference) but not on pointer manipulation.  Dereferencing
```

\note{I feel like this section wants a reference to CHERI Concentrate?}{nwf}

> ISO C permits pointers to go only one byte beyond their original

Author: I tried to use `address` when talking about the specific address component of a capability and use `pointer` to mean a C language pointer or the capability as a whole when used as a pointer.  I think some of the existing language here and elsewhere might use `pointer` to mean just the address?

> allocation, but real code often \psnote{not sure how strong to be

Reviewer: I would change "often" to "sometimes". If we look at the data I gathered from the FreeBSD test suite, it was 6 unique program counters per address space ID. At least one of those is memcpy(), so I wouldn't say very often.

Here's the data on OOB pointers from my dissertation/the POPL paper:
> We collected statistics for out-of-bounds pointer creation during a full run of the FreeBSD test suite. We encountered 155214 pointers that were out-of-bounds by more than one byte. Of these, only 1205 pointed past the end. This indicates that despite construction of one-past-the-end pointers being legal, having pointers to before the object is more common in real-world code. We also found that 81% of these pointers are at most sizeof(void*) beyond the bounds. While this is a surprising number of out-of-bounds pointers, there were at most six unique program counters generating these per process. One typical use of an out-of-bounds pointer is an idiom in which an array pointer is incremented prior to use within a loop, leading programmers to decrement the pointer below the lower bound in the loop. This idiom can be found in the zlib compression library, affecting dependent software such as gzip, OpenSSH, and libpng.
> While six unique program counters per program may seem like a negligible amount of code to be changed, every additional restriction imposed by pure-capability C/C++ on existing code increases the cost of adoption and therefore makes it less likely that CHERI could be deployed outside of an academic setting. Therefore, we conclude that the additional encoding bit used to represent out-of-bounds pointers is a good trade-off since it allows more source code to compile and run unmodified.

Author: Ok.  One thing the test suite doesn't cover are the weird things like in mandoc, though I suspect those are a small number of cases.

  there} constructs transient pointer values which are further out of
bounds.  To support this, capabilities can hold a range of
out-of-bounds addresses while retaining a valid tag.  Dereferencing an
out-of-bounds pointer will raise a hardware exception (see
\Cref{sec:faults}).  However, an out-of-bounds pointer can be
dereferenced once it has been brought back in bounds by adjusting the

> address or suppling a suitable offset in the dereference.  This means

Reviewer: Maybe it's slightly easier to understand for readers if we explicitly state when bounds checks are performed?
E.g. "CHERI-enabled hardware performs bounds checks on pointer use (i.e. dereference) and not pointer manipulation."

Author: Agreed.  I've taken a stab.

> that operations which modify the address of a capability do not clear
> the tag for out-of-bounds addresses.  Instead, a failure occurs if an

Reviewer: Maybe slightly easier to understand:
```suggestion
that operations which modify the address of a capability do not clear
the tag if the resulting address is out-of-bounds.  Instead, a failure occurs if an
```

Author: I do think this is clearer, but I've axed the sentence when adopting your earlier change.  If we bring it back we can add this fix to it.
